### PR TITLE
Refactor duplicate code for stringToNumber() in JSONObject, JSONArray, and XML

### LINKED ISSUE #813: Number parsing - de-duplication of logic
Logic to parse number is duplicated in XML and JSON parser.
They can be moved to single place and be referred by both.

### PR DESCRIPTION
**Pull request by rudrajyotib**
Both XML and JSON parser do convert string to number. Similar code has been duplicated for both parsers, which has been moved to a single utility method.

Fixes #813

### CODE REVIEW

File: src/main/java/org/json/NumberConversionUtil.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
+package org.json;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+public class NumberConversionUtil {
```
Comment (johnjaylward) [inline src/main/java/org/json/NumberConversionUtil.java:6]: Let's not make this public. We may wish to refactor this class in the future. For now make it private or default (package) visible.

**Review by johnjaylward — CHANGES_REQUESTED** [review]

File: src/main/java/org/json/NumberConversionUtil.java (hunk 1 above)
Comment (rudrajyotib) [reply src/main/java/org/json/NumberConversionUtil.java]: Static methods meant to be used by functional classes being marked private do not add up. But yes, @stleary - unnecessary making these classes open for used by clients might not be a great idea, there are plenty of scope of improvements in there. So marked them default visibility.

Comment (johnjaylward) [reply src/main/java/org/json/NumberConversionUtil.java]: That is correct. I simply meant that the class itself should be default or private, not the static methods.

Comment (johnjaylward) [reply src/main/java/org/json/NumberConversionUtil.java]: However, I now recall that private outer classes is not a thing in Java. I was thinking inner class scopes. Default visibility is correct here.

Comment (rudrajyotib) [reply src/main/java/org/json/NumberConversionUtil.java]: @stleary @johnjaylward - can this PR be accepted please?

Comment (stleary) [reply src/main/java/org/json/NumberConversionUtil.java]: @rudrajyotib Will get to it as soon as I can after the work of #811 is completed.

**Comment by rudrajyotib** [comment]
@stleary - can you accept this PR please?

**Comment by stleary** [comment]
@rudrajyotib Please update the PR:

From https://github.com/rudrajyotib/JSON-java
 * branch            issue813   -> FETCH_HEAD
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
fatal: Need to specify how to reconcile divergent branches.

File: src/test/java/org/json/NumberConversionUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    public void shouldParseMixedDecimalFractionNumbersWithoutLeadingZero(){
+        Number number = NumberConversionUtil.stringToNumber("200.10d");
+        assertEquals("Do not match",  200.10d, number.doubleValue(),0.0d);
+        assertEquals("Do not match",  200.10f, number.floatValue(),0.0f);
+        assertEquals("Do not match",  200, number.longValue(),0);
+        assertEquals("Do not match",  200, number.intValue(),0);
+    }
+
+
+     @Test
```
Comment (Mifrill) [inline src/test/java/org/json/NumberConversionUtilTest.java:59]: ⚙️ alignment
```suggestion
    @Test
```

**Comment by rudrajyotib** [comment]
@stleary - PR is now aligned with HEAD.

**Comment by rudrajyotib** [comment]
@stleary - can you please see if this can be accepted, or is it still showing some discrepancy across git branches.

**Comment by stleary** [comment]
**What problem does this code solve?**
Refactoring duplicate code

**Does the code still compile with Java6?**
Yes

**Risks**
Low

**Changes to the API?**
No

**Will this require a new release?**
No

**Should the documentation be updated?**
No

**Does it break the unit tests?**
No, new unit tests were added

**Was any code refactored in this commit?**
Yes, this was a refactoring ticket

**Review status**
**APPROVED - Pending completion of code review**

Start 3-day comment window

File: src/main/java/org/json/JSONObject.java [hunk 3]
```diff
@@ -28,6 +28,9 @@
 import java.util.Set;
 import java.util.regex.Pattern;
 
+import static org.json.NumberConversionUtil.potentialNumber;
```
Comment (stleary) [inline src/main/java/org/json/JSONObject.java:31]: Just import NumberConversionUtil, not the methods. Someone glancing at the code might not realize it is a static method in another class.

File: src/main/java/org/json/NumberConversionUtil.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     *
+     * @param val value to test
+     * @return true if the string is "-0" or if it contains '.', 'e', or 'E', false otherwise.
+     */
+    private static boolean isDecimalNotation(final String val) {
+        return val.indexOf('.') > -1 || val.indexOf('e') > -1
+                || val.indexOf('E') > -1 || "-0".equals(val);
+    }
+
+    private static boolean potentialPositiveNumberStartingAtIndex(String value,int index){
```
Comment (stleary) [inline src/main/java/org/json/NumberConversionUtil.java:109]: There is a duplicate of this method in JSONObject. It should be removed, and the calling code updated to use the NumberConversionUtil method.

File: src/main/java/org/json/XML.java [hunk 5]
```diff
@@ -10,6 +10,9 @@
 import java.math.BigInteger;
 import java.util.Iterator;
 
+import static org.json.NumberConversionUtil.potentialNumber;
```
Comment (stleary) [inline src/main/java/org/json/XML.java:13]: Just import NumberConversionUtil, not the methods.

File: src/main/java/org/json/XML.java [hunk 6]
```diff
@@ -495,104 +498,9 @@ public static Object stringToValue(String string) {
...
-                            throw new NumberFormatException("val ["+input+"] is not a valid number.");
-                        }
-                        return d;
-                    } catch (NumberFormatException ignore) {
-                        throw new NumberFormatException("val ["+input+"] is not a valid number.");
-                    }
-                }
-            }
-            val = removeLeadingZerosOfNumber(input);
-            if(initial == '0' && val.length() > 1) {
```
Comment (stleary) [inline src/main/java/org/json/XML.java:557]: The JSONObject version of this method inserts this line, which is missing from the XML copy:
`initial = val.charAt(0);`. I don't think it matters, though.

**Comment by stleary** [comment]
@rudrajyotib  Looks good to me. I left some comments in the code. They are all trivial issues that can wait until this commit is merged. Please do not make any more changes to this PR; if you do, the accept status will be taken down and it won't get committed before the end of Hacktoberfest. I will open an issue to address the notes after Nov 1, but they are low priority.

**Comment by rudrajyotib** [comment]
@stleary sure thing. I will update them in another PR.

File: src/main/java/org/json/XML.java (hunk 6 above)
Comment (rudrajyotib) [reply src/main/java/org/json/XML.java:557]: This line is not needed in any of the versions, and has been deleted from either, moved into the utility static method.

File: src/main/java/org/json/XML.java (hunk 5 above)
Comment (rudrajyotib) [reply src/main/java/org/json/XML.java:13]: In the next PR it would be changed.

File: src/main/java/org/json/NumberConversionUtil.java (hunk 4 above)
Comment (rudrajyotib) [reply src/main/java/org/json/NumberConversionUtil.java:109]: Right, the duplicate method will be removed in the next PR.

**Review by stleary — APPROVED** [review]

**Pull request merged**